lexer/impl/lexer.cpp

```diff
@@ -85,7 +85,10 @@ Token Lexer::Peek() {
     if (currentToken == nullptr) {
         currentToken = std::make_shared<Token>(scanNext());
     } else if (currentToken->type == TokenType::Eof) {
-        throw "End-of-file already reached";
+        Token t;
+        t.type = TokenType::Eof;
+        return t;
+        // throw "End-of-file already reached";
```

Comment:
Right, but I had already implemented the same functionality before you pushed, so I didn't merge it. Also, there are a few points I didn't like about that implementation; mainly that it calls `Next()` in the constructor, meaning that it isn't really lazy evaluation and that calling `Next()` on the first actual usage will discard the first token (requiring to use `Peek` on the first call to the lexer). My implementation provides a predictable behavior where you can use either `Next` or `Peek` in any order you want, and only when you call `Next` would the next token be actually consumed.

About EOF, I am not saying we shouldn't return the EOF token, because we do. All I am saying is that after we have already reached the end-of-file **and** consumed that token (using `Next` instead of `Peek`), there is no next token to consume or peek at, so it doesn't make sense to return EOF again and it should be considered an error